### PR TITLE
ci(kitchen+gitlab-ci): use latest pre-salted images (inc. fixes and tests)

### PR DESCRIPTION
**Pull request by myii**
### PR progress checklist (to be filled in by reviewers)

- [ ] Changes to documentation are appropriate (or tick if not required)
- [ ] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

#### Primary type

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [x] `[ci]`       Changes to the continuous integration configuration
- [ ] `[feat]`     A new feature
- [x] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [ ] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [ ] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type

- [ ] `[docs]`     Documentation changes
- [x] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?

No.

### Related issues and/or pull requests

* Semi-automated using https://github.com/myii/ssf-formula/pull/293

### Describe the changes you're proposing

Usually when updating to the latest pre-salted images, there are minimal changes or just tests that need to be updated -- these are done without submitting a PR.  However, this was far more extensive here.

The changes proposed here allow all platforms to be enabled in the CI, except for Gentoo.

---

#### Commit: fix(templates/haproxy.jinja): replace deprecated `reqadd`

https://github.com/haproxy/haproxy/blob/31dd393da0e6c20bf65ea833d10635a8b26cb355/src/cfgparse-listen.c#L2843-L2845

```c
    else if (strcmp(args[0], "reqadd") == 0) {  /* add request header */
           ha_alert("parsing [%s:%d] : The '%s' directive is not supported anymore since HAProxy 2.1. "
                    "Use 'http-request add-header' instead.\n", file, linenum, args[0]);
```

Introducing this without a breaking change by automatically using `http-request add-header` instead of `reqadd`.

---

#### Commit: fix(config): ensure `haproxy:global:chroot:path` is created if provided

This path isn't always available after installation (e.g. on Arch Linux), so ensure it is created.  Fixed the one-space indent in the state above as well.

---

#### Commit: test(pillar): provide `default` pillar working on all platforms

Attempted to keep the test pillar as close to `pillar.example` as possible (which was the previous test pillar).

Here's the diff:

```diff
--- .../haproxy-formula/pillar.example
+++ .../haproxy-formula/test/salt/pillar/default.sls
@@ -1,8 +1,4 @@
-#
-# Example pillar configuration
-#
 ---
-
 haproxy:
   # use lookup section to override 'map.jinja' values
   # lookup:
@@ -26,6 +22,7 @@
     log-send-hostname: localhost
     stats:
       enable: true
+      # Using the `haproxy:global:chroot:path`
       socketpath: /var/lib/haproxy/stats
       mode: 660
       level: admin
@@ -76,23 +73,6 @@
       - realm: 'Haproxy\ Statistics'
       - auth: 'admin1:AdMiN123'
 
-    errorfiles:
-      400: /etc/haproxy/errors/400.http
-      403: /etc/haproxy/errors/403.http
-      408: /etc/haproxy/errors/408.http
-      500: /etc/haproxy/errors/500.http
-      502: /etc/haproxy/errors/502.http
-      503: /etc/haproxy/errors/503.http
-      504: /etc/haproxy/errors/504.http
-
-  resolvers:
-    local_dns:
-      options:
-        - nameserver resolvconf 127.0.0.1:53
-        - resolve_retries 3
-        - timeout retry 1s
-        - hold valid 10s
-
 
   listens:
     stats:
@@ -136,7 +116,6 @@
       tcpchecks:
         - send PINGrn
         - expect string +PONG
-        - send info replicationrn
         - expect string role:master
         - send QUITrn
         - expect string +OK
@@ -158,7 +137,7 @@
       redirects:
         - scheme https if !{ ssl_fc }
       reqadds:
-        - "X-Forwarded-Proto:\\ http"
+        - "X-Forwarded-Proto http"
       default_backend: www-backend
 
   # www-https:
@@ -166,7 +145,7 @@
   # yamllint disable-line rule:line-length
   #   logformat: "%ci:%cp\\ [%t]\\ %ft\\ %b/%s\\ %Tq/%Tw/%Tc/%Tr/%Tt\\ %ST\\ %B\\ %CC\\ %CS\\ %tsc\\ %ac/%fc/%bc/%sc/%rc\\ %sq/%bq\\ %hr\\ %hs\\ %{+Q}r\\ ssl_version:%sslv\\ ssl_cipher:%sslc"
   #   reqadds:
-  #        - "X-Forwarded-Proto:\\ https"
+  #        - "X-Forwarded-Proto https"
   #      default_backend: www-backend
   #      acls:
   #        - url_static       path_beg       -i /static /images /javascript /stylesheets
@@ -184,9 +163,9 @@
     backend1:
       name: www-backend
       balance: roundrobin
+      extra: "http-request del-header ^X-Forwarded-For:"
       redirects:
         - scheme https if !{ ssl_fc }
-      extra: "reqidel ^X-Forwarded-For:"
       servers:
         server1:
           name: server1-its-name
@@ -224,26 +203,3 @@
           host: apiserver1.example.com
           port: 80
           check: check
-        server2:
-          name: apiserver2
-          host: apiserver2.example.com
-          port: 80
-          check: check
-          extra: resolvers local_dns resolve-prefer ipv4
-    another_www:
-      mode: tcp
-      balance: source
-      sticktable: "type binary len 32 size 30k expire 30m"
-      acls:
-        - clienthello req_ssl_hello_type 1
-        - serverhello rep_ssl_hello_type 2
-      tcprequests:
-        - "inspect-delay 5s"
-        - "content accept if clienthello"
-      tcpresponses:
-        - "content accept if serverhello"
-      stickons:
-        - "payload_lv(43,1) if clienthello"
-      reqreps:
-        - '^([^\ :]*)\ /static/(.*)	\1\ \2'
-      options: "ssl-hello-chk"
```

---

### Documentation checklist

- [ ] Updated the `README` (e.g. `Available states`).
- [ ] Updated `pillar.example`.

### Testing checklist

- [x] Included in Kitchen (i.e. under `state_top`).
- [x] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [x] Updated the relevant test pillar.

### CODE REVIEW

File: haproxy/config.sls [hunk 1]
```diff
@@ -2,18 +2,30 @@
...
+    - group: {{ haproxy.group }}
+    - mode: 644
+    - require_in:
+      - service: haproxy.service
+    - watch_in:
+      - service: haproxy.service
+    {% if salt['pillar.get']('haproxy:overwrite', default=True) == False %}
+    - unless:
+      - test -e {{ config_file }}
+    {% endif %}
```
Comment (myii) [inline haproxy/config.sls:19]: No change other than fixing the indent.

File: haproxy/config.sls [hunk 2]
```diff
@@ -2,18 +2,30 @@
...
+      salt['pillar.get']('haproxy:global:chroot:path', '') %}
+haproxy-chroot-directory:
+  file.directory:
+    - name: {{ salt['pillar.get']('haproxy:global:chroot:path') }}
+    - user: {{ haproxy.user }}
+    - group: {{ haproxy.group }}
+    - dir_mode: 755
+    - require_in:
+      - service: haproxy.service
+{% endif %}
```
Comment (myii) [inline haproxy/config.sls:31]: #### Commit: fix(config): ensure `haproxy:global:chroot:path` is created if provided

File: haproxy/templates/haproxy.jinja [hunk 3]
```diff
@@ -282,10 +282,10 @@ listen {{ listener.get('name', listener_name) }}
     {%- endif %}
     {%- if 'reqadds' in listener %}
       {%- if listener.reqadds is string %}
-    reqadd {{ listener.reqadds }}
+    http-request add-header {{ listener.reqadds }}
       {%- else %}
         {%- for reqadd in listener.reqadds %}
-    reqadd {{ reqadd }}
+    http-request add-header {{ reqadd }}
```
Comment (myii) [inline haproxy/templates/haproxy.jinja:288]: #### Commit: fix(templates/haproxy.jinja): replace deprecated `reqadd`

File: haproxy/templates/haproxy.jinja [hunk 4]
```diff
@@ -450,7 +450,7 @@ frontend {{ frontend.get('name', frontend_name) }}
     {{- render_list_of_dictionaries('reqrep', frontend.reqreps) }}
     {%- endif %}
     {%- if 'reqadds' in frontend %}
-    {{- render_list_of_dictionaries('reqadd', frontend.reqadds) }}
+    {{- render_list_of_dictionaries('http-request add-header', frontend.reqadds) }}
```
Comment (myii) [inline haproxy/templates/haproxy.jinja:453]: #### Commit: fix(templates/haproxy.jinja): replace deprecated `reqadd`

File: test/salt/pillar/default.sls [hunk 5]
```diff
@@ -0,0 +1,205 @@
+---
```
Comment (myii) [inline test/salt/pillar/default.sls:1]: #### Commit: test(pillar): provide `default` pillar working on all platforms

**Review by myii — COMMENTED** [review]
@daks In order to make reviewing easier, I've added a comment to each portion that needs looking at.  All of the rest of the files are standard changes made by the `ssf-formula` and can be ignored.

File: haproxy/config.sls (hunk 2 above)
Comment (daks) [reply haproxy/config.sls:31]: I don't see any default value for this path on mapdata, does this `file.directory` works well if `name` is `''`?

**Review by daks — APPROVED** [review]
Except for one comment, all seems fine for me and can be merged!

File: haproxy/config.sls (hunk 2 above)
Comment (myii) [reply haproxy/config.sls:31]: > I don't see any default value for this path on mapdata, does this `file.directory` works well if `name` is `''`?

If `name` is `''`, it won't even render the state due to the second condition of the `if` block it's within:

```jinja
{% if ... and
      salt['pillar.get']('haproxy:global:chroot:path', '') %}
```

**Comment by myii** [comment]
Really appreciate the review, @daks.  Another formula converted to the latest pre-salted images (well, not quite, we pushed the very latest yesterday!)

**Pull request merged**